File: codes/go/operator_optional_chaining/operator_optional_chaining.go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
)

const EMPTY_STRING = ""
const TAB = "    "

// type any interface{}
type array []any
type object map[string]any

func prettyJsonStringify(anything any) string {
	marshalledJson, err := json.MarshalIndent(anything, EMPTY_STRING, TAB)
	if (err == nil) {
		return string(marshalledJson)
	}

	return "undefined"
}

func arrayReduce(callbackFunction func(any, any, int, array) any, anArray array, initialValue any) any {
	// JavaScript-like Array.reduce() function
	result := initialValue
	for arrayItemIndex, arrayItem := range anArray {
		result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
	}
	return result
}

func optionalChaining(anything any, objectPropertiesArray ...any) any {
	anythingType := reflect.TypeOf(anything).Kind()
	if (((anythingType != reflect.Map) && (anythingType != reflect.Slice)) || (len(objectPropertiesArray) == 0)) {
		return anything
	}
	return arrayReduce(func(currentResult any, currentItem any, _ int, _ array) any {
		if (currentResult == nil) {
			return anything.(object)[currentItem.(string)]
		}
		currentResultType := reflect.TypeOf(currentResult).Kind()
		if (currentResultType == reflect.Map) {
			return currentResult.(object)[currentItem.(string)]
		}
		if ((currentItem.(int) >= 0) && (currentItem.(int) < len(currentResult.(array)))) {
			return currentResult.(array)[currentItem.(int)]
		}
		return nil
	}, objectPropertiesArray, nil)
}

func main() {
	fmt.Println("\n// JavaScript-like Optional Chaining Operator (?.) in Go")

	// There's no JavaScript-like Optional Chaining Operator (?.) in Go.
	// But, we can create our own function to mimic it in Go.

	JSON_OBJECT := object{
		"foo": object{
			"bar": "baz",
		},
		"fruits": array{"apple", "mango", "banana"},
	}
	fmt.Println("JSON_OBJECT:", prettyJsonStringify(JSON_OBJECT))

	fmt.Println("JSON_OBJECT?.foo?.bar:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "foo", "bar")))
	// JSON_OBJECT?.foo?.bar: "baz"

	fmt.Println("JSON_OBJECT?.foo?.baz:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "foo", "baz")))
	// JSON_OBJECT?.foo?.baz: null

	fmt.Println("JSON_OBJECT?.fruits?.[2]:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "fruits", 2)))
	// JSON_OBJECT?.fruits?.[2]: "banana"

	fmt.Println("JSON_OBJECT?.fruits?.[5]:", prettyJsonStringify(optionalChaining(JSON_OBJECT, "fruits", 5)))
	// JSON_OBJECT?.fruits?.[5]: null
}

File: codes_trial/matlab/operator_optional_chaining.m
function println(varargin)
    result = {};
    for parameterIndex = 1:1:length(varargin)
        if isa(varargin{parameterIndex}, "double")
            result = [result, num2str(varargin{parameterIndex})];
            continue
        end
        if (varargin{parameterIndex} == true)
            result = [result, "true"];
            continue
        end
        if (varargin{parameterIndex} == false)
            result = [result, "false"];
            continue
        end
        result = [result, varargin{parameterIndex}];
    end
    result = [result{:}];
    disp(result);
end

function prettyJsonStringifyResult = prettyJsonStringify(parameter)
    if ((nargin == 0) || (isa(parameter, "cell") == false))
        prettyJsonStringifyResult = "";
        return
    end
    indent = "    ";
    if ((nargin == 2) && (isa(parameter{2}, "struct") == true) && isfield(parameter{2}, "indent") == true)
        indent = parameter.indent;
    end
    anything = parameter{1};
    indentLevel = 0;
    function prettyJsonStringifyInnerFunctionResult = prettyJsonStringifyInnerFunction(anything, indent)
        if (anything == "null")
            prettyJsonStringifyInnerFunctionResult = "null";
            return
        end
        if (anything == "undefined")
            prettyJsonStringifyInnerFunctionResult = "undefined";
            return
        end
        if (isa(anything, "char") == true)
            prettyJsonStringifyInnerFunctionResult = ["""", anything, """"];
            return
        end
        if (isa(anything, "double") == true)
            prettyJsonStringifyInnerFunctionResult = num2str(anything);
            return
        end
        if (anything == true)
            prettyJsonStringifyInnerFunctionResult = "true";
            return
        end
        if (anything == false)
            prettyJsonStringifyInnerFunctionResult = "false";
            return
        end
        if (isa(anything, "cell") == true)
            prettyJsonStringifyInnerFunctionResult = "cell";
            return
        end
        if (isa(anything, "struct") == true)
            prettyJsonStringifyInnerFunctionResult = "struct";
            return
        end
        prettyJsonStringifyInnerFunctionResult = "undefined";
    end
    prettyJsonStringifyResult = prettyJsonStringifyInnerFunction(anything, indent);
end
% asd = prettyJsonStringify({"qwe", 2});
% println("hello ", asd);
% println(asd);
% println(class({1, 2, 3}));
% disp(class("asd"));

println("\n\% JavaScript-like Optional Chaining Operator (?.) in MATLAB");

% There's no JavaScript-like Optional Chaining Operator (?.) in MATLAB.
% But, we can create our own function to mimic it in MATLAB.

JSON_OBJECT = { ...
    struct( ...
        "foo", struct( ...
            "bar", "baz" ...
        ) ...
    ), ...
    struct( ...
        "fruits", {"apple", "mango", "banana"} ...
    ) ...
};
